### PR TITLE
feat(firestore): start withConverter feature

### PR DESCRIPTION
**Pull request by Ehesp**
### Description

Adds in the `withConverter` functionality to Firestore.

### Related issues

https://invertase.canny.io/admin/board/react-native-firebase/p/support-firestore-dataconverter-withconverter

### Checklist

- I read the [Contributor Guide](/CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [x] Yes
- My change supports the following platforms;
  - [x] `Android`
  - [x] `iOS`
- My change includes tests;
  - [ ] `e2e` tests added or updated in `packages/\*\*/e2e`
  - [ ] `jest` tests added or updated in `packages/\*\*/__tests__`
- [ ] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [x] No

### Test Plan

Need to test both E2E, unit tests & also add in some TypeScript tests, I'm not 100% sure yet how the types filter down to the returned documents... eg:

```
const collection = firestore.collection('foo').withConverter<ExpectedResult>({});

const docSnapshot = await collection.get();

// Should return type of "docSnapshot.data()" now be ExpectedResult?

```

### CODE REVIEW

File: packages/firestore/lib/index.d.ts [hunk 1]
```diff
@@ -1349,7 +1358,7 @@ export namespace FirebaseFirestoreTypes {
      * Changes the behavior of `set()` calls to only replace the specified field paths.
      * Any field path that is not specified is ignored and remains untouched.
      */
-    mergeFields?: (string | FieldPath)[];
+    mergeFields?: string | FieldPath[];
```
Comment (Salakar) [inline packages/firestore/lib/index.d.ts:1361]: Shouldn't this still be an array of either string or field path, this is no longer an array after this change

File: packages/firestore/lib/utils/index.js [hunk 2]
```diff
@@ -233,3 +233,22 @@ export function parseSnapshotArgs(args) {
...
   return { snapshotListenOptions, callback, onNext, onError };
 }
+
+/**
+ * Validates a withConverter object contains both required functions
+ *
+ * @param converter
+ */
+export function validateWithConverter(converter) {
+  if (!isUndefined(converter) || !isObject(converter)) {
```
Comment (Salakar) [inline packages/firestore/lib/utils/index.js:243]: !undefined seems wrong

File: packages/firestore/lib/utils/index.js [hunk 3]
```diff
@@ -233,3 +233,22 @@ export function parseSnapshotArgs(args) {
...
+    throw new Error('expected an object value.');
+  }
+
+  if (!converter.toFirestore || !isFunction(converter.toFirestore)) {
```
Comment (Salakar) [inline packages/firestore/lib/utils/index.js:247]: !converter unnecessary, isFunction check will catch it anyway

File: packages/firestore/lib/utils/index.js [hunk 4]
```diff
@@ -233,3 +233,22 @@ export function parseSnapshotArgs(args) {
...
+    throw new Error("'toFirestore' expected a function.");
+  }
+
+  if (!converter.fromFirestore || !isFunction(converter.fromFirestore)) {
```
Comment (Salakar) [inline packages/firestore/lib/utils/index.js:251]: !converter unnecessary, isFunction check will catch it anyway

**Comment by Salakar** [comment]
DocumentSnapshot type would  need changing to add in T,

E.g.

```ts
interface DocumentSnapshot<T = DocumentSnapshotDataType> {
  // .. 
  data(): T;
}
```

DocumentSnapshotDataType being whatever it currently returns without the converter changes

File: packages/firestore/lib/index.d.ts (hunk 1 above)
Comment (Ehesp) [reply packages/firestore/lib/index.d.ts:1361]: I think that was prettier actually, didn't manually change it. I guess correct types are `string[] | FieldPath[]`

**Comment by sperezm97** [comment]
Hey guys ! Any update ? if you need help, im here ! thanks

File: packages/firestore/lib/index.d.ts [hunk 5]
```diff
@@ -461,6 +461,15 @@ export namespace FirebaseFirestoreTypes {
      * @param moreFieldsAndValues Additional key value pairs.
      */
     update(field: string | FieldPath, value: any, ...moreFieldsAndValues: any[]): Promise<void>;
+
+    withConverter<U>(converter: FirestoreDataConverter<U>): DocumentReference<U>;
+  }
+
+  // TODO test this
```
Comment (dackers86) [inline packages/firestore/lib/index.d.ts:468]: Todo?

File: packages/firestore/lib/FirestoreDocumentSnapshot.js [hunk 6]
```diff
@@ -72,6 +73,11 @@ export default class FirestoreDocumentSnapshot {
     //   }
```
Comment (dackers86) [inline packages/firestore/lib/FirestoreDocumentSnapshot.js:73]: Can this be removed?

File: packages/firestore/lib/index.d.ts [hunk 7]
```diff
@@ -1645,6 +1654,9 @@ export namespace FirebaseFirestoreTypes {
       value: any,
       ...moreFieldsAndValues: any[]
     ): Transaction;
+
+    // TODO test passing U down? How does it effect the documents
```
Comment (dackers86) [inline packages/firestore/lib/index.d.ts:1658]: Todo?

**Comment by andrefedev** [comment]
I am migrating my application to react native and I am using this package, I am concerned that such a necessary function in the Typescript world has not been implemented yet. Any other alternative to handle this?

**Comment by mikehardy** [comment]
There are no plans @andrefedev - it's a very limited crew (basically, me, very part time) where I have the main task of shepherding community contributions to successful testing and release so that the package moves forward effectively but only so far as the community contributes. 

As such I would be *more* than happy if you picked up this PR and brought it to completion, and I would assist in doing so, but someone from the community with sufficient interest would have to be the prime mover on it, otherwise we infer that it's not important enough and it sits. 

The original author - Eliot - is booked on other things as I understand it - so picking it up would be welcome to everyone

**Comment by anhnch** [comment]
I can't wait for this feature

**Comment by mikehardy** [comment]
@anhnch PRs gladly accepted - you may use this as a basis for work, just let us know and we'll coordinate for merge/release

**Comment by dogansalman** [comment]
> I can't wait for this feature

+1

**Comment by JanErikFoss** [comment]
What's the status here?

**Comment by mikehardy** [comment]
All development here is out in the open, the comments you see are the status there is

**Comment by tajetaje** [comment]
What is needed to get this merged? To get this going is it just tests and updating it with main? Or is there functionality missing?

**Comment by mikehardy** [comment]
I'm unsure actually - firestore files have not changed too drastically so the merge conflict is hopefully not too bad to resolve.
@Ehesp has been poking around the repo the last couple days, he may have current thoughts on this?

**Comment by superphil0** [comment]
Is this still in the making or did this get dropped from the roadmap?

**Comment by mikehardy** [comment]
It's an exciting feature, but the current roadmap is - "finish firebase-js-sdk v9 adoption", "typescript", "support react-native-web (somehow)". This feature may make it's way inside that list somewhere but that list is where we are focusing at the moment

**Comment by ChromeQ** [comment]
I would hope that this feature falls under the "typescript" project on the roadmap @mikehardy 
Converting a project from web to native means it isn't as straightforward as I hoped on the firestore stuff, looking forward to seeing more on this making progress.
Is the "solution" for now to cast `as Type`? And also manually call the conversion functions (Timestamp to Date etc)?

**Comment by jthoward64** [comment]
> I would hope that this feature falls under the "typescript" project on the roadmap @mikehardy Converting a project from web to native means it isn't as straightforward as I hoped on the firestore stuff, looking forward to seeing more on this making progress. Is the "solution" for now to cast `as Type`? And also manually call the conversion functions (Timestamp to Date etc)?

That or wrap the calls to firestore and use `as unknown` and then throw it through a type check function

**Comment by haeniya** [comment]
@ChromeQ How did you go about this? Was it possible to just call the conversion functions manually?

**Comment by ChromeQ** [comment]
@haeniya I just used a TS cast to the type I know it should be (although it should be safer to do as @jthoward64 suggests to have a runtime check with a type check function)

```ts
const snapshot = await firestore.collection('days').doc(day).get();
const { pins } = snapshot.data() as { pins: DocumentReference<FromFirestorePin>[] };
```

I have defined two types which represent the differences between a front-end ready json object and the actual return from firestore:
```ts
export interface ToFirestorePin extends Omit<Pin, 'updatedAt'> {
  updatedAt: FirebaseFirestoreTypes.FieldValue;
}

export interface FromFirestorePin extends Omit<Pin, 'updatedAt'> {
  updatedAt: FirebaseFirestoreTypes.Timestamp;
}
```

And then I also have a "conversion" function like this:
```ts
export const convertFirestorePinToPin = (
  doc:
    | FirebaseFirestoreTypes.QueryDocumentSnapshot<FromFirestorePin>
    | FirebaseFirestoreTypes.DocumentSnapshot<FromFirestorePin>
): Pin => {
  const data = doc.data();

  if (!data) {
    throw new Error(`No data found for firestore document`);
  }

  return {
    ...data,
    id: data.id,
    updatedAt: data.updatedAt.toDate().toISOString(),
  };
};
```

You can just call this:
```ts
const snapshot = await firestore.collection<FromFirestorePin>('pins').doc(pinId).get();
const pin = convertFirestorePinToPin(snapshot);
```

Hope this helps and I am only converting the Timestamp to a UTC ISO string but there are a few other differences that might trip you up. Good luck

**Comment by haeniya** [comment]
@ChromeQ many thanks! That should get me started.

**Pull request closed without merging**